File: models/dislike.ts
import { type DocumentData, type DocumentReference, Timestamp } from 'firebase/firestore'

export interface IDislike {
  whoDisliked: DocumentReference | null
  dislikedProfile: DocumentReference | null
  date: Timestamp | Date
}

export class DislikeModel implements IDislike {
  whoDisliked: DocumentReference | null
  dislikedProfile: DocumentReference | null
  date: Timestamp | Date

  reference: DocumentReference | null

  constructor(data: IDislike, reference: DocumentReference | null) {
    this.whoDisliked = data.whoDisliked || null
    this.dislikedProfile = data.dislikedProfile || null
    this.date = data.date instanceof Timestamp ? data.date.toDate() : data.date

    this.reference = reference
  }

  toMap() {
    return {
      whoDisliked: this.whoDisliked,
      dislikedProfile: this.dislikedProfile,
      date: this.date,
    }
  }
}

export function mapDislike(like: DocumentData) {
  return new DislikeModel(
    like.data(),
    like.ref,
  )
}

File: functions/src/models/dislike.py
import dataclasses
import datetime
import inspect
import typing

from google.cloud import firestore


@dataclasses.dataclass(kw_only=True, frozen=True, repr=False)
class DislikeModel:
    whoDisliked: typing.Optional[firestore.DocumentReference] = None
    dislikedProfile: typing.Optional[firestore.DocumentReference] = None
    date: datetime.datetime = datetime.datetime.now()

    reference: typing.Optional[firestore.DocumentReference] = None

    def to_map(self):
        data = dataclasses.asdict(self)
        del data["reference"]

        return data

    @classmethod
    def from_dict(cls, data):
        return cls(
            **{
                key: value
                for key, value in data.items()
                if key in inspect.signature(cls).parameters
            }
        )
